src/reconcile/rec_visibility.c

```diff
@@ -758,3 +758,253 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
+            upd_select->stop_ts = WT_TS_MAX;
+            upd_select->stop_txn = WT_TXN_MAX;
+        }
+    }
+
+    /*
+     * Track the most recent transaction in the page. We store this in the tree at the end of
+     * reconciliation in the service of checkpoints, it is used to avoid discarding trees from
+     * memory when they have changes required to satisfy a snapshot read.
+     */
+    if (WT_TXNID_LT(r->max_txn, max_txn))
+        r->max_txn = max_txn;
+
+    /* Update the maximum timestamp. */
+    if (max_ts > r->max_ts)
+        r->max_ts = max_ts;
+
+    /*
+     * If the update we chose was a birthmark.
+     */
+    if (upd != NULL && upd->type == WT_UPDATE_BIRTHMARK) {
+        /*
+         * Resolve the birthmark now regardless of whether the update being written to the data file
+         * is the same as it was the previous reconciliation. Otherwise lookaside can end up with
+         * two birthmark records in the same update chain.
+         */
+        WT_RET(__rec_append_orig_value(session, page, first_upd, vpack));
+        upd_select->upd = NULL;
+    }
+
+    /*
+     * Check if all updates on the page are visible, if not, it must stay dirty.
+     *
+     * Updates can be out of transaction ID order (but not out of timestamp order), so we track the
+     * maximum transaction ID and the newest update with a timestamp (if any).
+     */
+    all_stable =
+      !list_prepared && !list_uncommitted && __wt_txn_visible_all(session, max_txn, max_ts);
+
+    if (all_stable)
+        goto check_original_value;
+
+    r->leave_dirty = true;
+
+    if (F_ISSET(r, WT_REC_VISIBILITY_ERR))
+        WT_PANIC_RET(session, EINVAL, "reconciliation error, update not visible");
+
+    /* If not trying to evict the page, we know what we'll write and we're done. */
+    if (!F_ISSET(r, WT_REC_EVICT))
+        goto check_original_value;
+
+    /*
+     * We are attempting eviction with changes that are not yet stable (i.e. globally visible).
+     * There are two ways to continue, the save/restore eviction path or the lookaside table
+     * eviction path. Both cannot be configured because the paths track different information. The
+     * update/restore path can handle uncommitted changes, by evicting most of the page and then
+     * creating a new, smaller page to which we re-attach those changes. Lookaside eviction writes
+     * changes into the lookaside table and restores them on demand if and when the page is read
+     * back into memory.
+     *
+     * Both paths are configured outside of reconciliation: the save/restore path is the
+     * WT_REC_UPDATE_RESTORE flag, the lookaside table path is the WT_REC_LOOKASIDE flag.
+     */
+    if (!F_ISSET(r, WT_REC_LOOKASIDE | WT_REC_UPDATE_RESTORE))
+        return (__wt_set_return(session, EBUSY));
+    if (list_uncommitted && !F_ISSET(r, WT_REC_UPDATE_RESTORE))
+        return (__wt_set_return(session, EBUSY));
+
+check_original_value:
```

Comment:
May be a better name. Here we should save updates for both eviction and checkpoint